File: problems/problems_698/solution.rs
use serde_json::{json, Value};

pub struct Solution;

impl Solution {
    pub fn can_partition_k_subsets(nums: Vec<i32>, k: i32) -> bool {
		let sum: i32 = nums.iter().sum();
		if sum % k != 0 {
			return false;
		}
		let target = sum / k;
		for num in nums.iter() {
			if *num > target {
				return false;
			}
		}
		let n = nums.len();
		let mut dp = vec![-1; 1 << n];
		dp[0] = 0;
		for mask in 0..1<<n {
			for i in 0..n {
				if (mask >> i) & 1 != 0 {
					let before = mask ^ (1 << i);
					if dp[before] != -1 && dp[before] + nums[i] <= target {
						dp[mask] = (dp[before] + nums[i]) % target;
					}
				}
			}
		}
		dp[(1 << n) - 1] == 0
    }
}

#[cfg(feature = "solution_698")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let nums: Vec<i32> = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	let k: i32 = serde_json::from_str(&input_values[1]).expect("Failed to parse input");
	json!(Solution::can_partition_k_subsets(nums, k))
}

File: problems/problems_698/solution.py
import solution
from typing import *
from functools import lru_cache


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.canPartitionKSubsets(*test_input)

    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        total = sum(nums)
        if total % k != 0:
            return False
        line = total // k
        if any(num > line for num in nums):
            return False
        n = len(nums)
        all_picked = (1 << n) - 1

        @lru_cache(None)
        def dfs(state, cur):
            if cur == line:
                if state == all_picked:
                    return True
                cur = 0
            for i in range(n):
                if not (state >> i) & 1 and cur + nums[i] <= line and dfs(state | (1 << i), cur + nums[i]):
                    return True
            return False

        return dfs(0, 0)
